### ISSUE TITLE
Fix plugin exception reporting keyerror typo - [merged]

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#908]
In GitLab by @Tadaboody on Dec 31, 2018, 13:08

_Merges error_plugin_exception_report -> master_

This made it crash and raise a key error before reporting the plugin exception.
I think there should be a non-depricated api before adding tests for this

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Dec 31, 2018, 13:12

What API is deprecated? Why would that affect whether a test should be added for this?

**Comment by asottile** [comment]
In GitLab by @asottile on Dec 31, 2018, 13:13

approved this merge request

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Dec 31, 2018, 13:14

The [public python API](http://flake8.pycqa.org/en/latest/user/python-api.html). I'd use it to write tests for a plugin

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Dec 31, 2018, 13:21

Except that we have [unit tests](https://gitlab.com/pycqa/flake8/blob/master/tests/unit/test_file_checker.py) for parts of this module anyway. You could, if you so desired, write a test for this method there. `plugin` is a dictionary and the `plugin` key (`plugin["plugin"]`) only needs a callable that would raise any possible exception to verify that this now raises the `PluginExecutionFailed` exception. Something like

```py
def test_raises_proper_exception_on_failed_plugin_run():
    fchecker = checker.FileChecker(...)
    plugin = {
        "parameters": [],
        "plugin": mock.MagicMock(side_effect=ValueError),
    }
    with pytest.raises(flake8.exceptions.PluginExecutionFailed):
        fchecker.run_check(plugin)
```

There's absolutely no need to use the legacy API

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Dec 31, 2018, 13:25

My bad then - would you like me to write tests for this MR?

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jan 1, 2019, 06:24

Yes, please write at least one test for this. You can even mostly copy-paste my example test if that helps  :smile:

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 12, 2019, 13:45

Hi, sorry I haven't pushed anything yet but I can't quite get the test to pass.
I've written essentially the same test as you (just filled the `...` and I get this error
```py
self = FileChecker for foobar.py, plugin = {'parameters': [{'ast': True}], 'plugin': <MagicMock id='4490204328'>}, arguments = {}

    def run_check(self, plugin, **arguments):
        """Run the check in a single plugin."""
        LOG.debug("Running %r with %r", plugin, arguments)
        try:
>           self.processor.keyword_arguments_for(
                plugin["parameters"], arguments
            )
E           AttributeError: 'NoneType' object has no attribute 'keyword_arguments_for'

src/flake8/checker.py:433: AttributeError
```
with the follow up
```py
During handling of the above exception, another exception occurred:

    def test_raises_proper_exception_on_failed_plugin_run():
        plugin = {
            "parameters": [{"ast":True}],
            "plugin": mock.MagicMock(side_effect=ValueError),
        }
        """Verify a failing plugin results in an plugin error"""
        fchecker = checker.FileChecker("foobar.py",checks={},options=object())
        with pytest.raises(flake8.exceptions.PluginExecutionFailed):
>           fchecker.run_check(plugin)

tests/unit/test_file_checker.py:57:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = FileChecker for foobar.py, plugin = {'parameters': [{'ast': True}], 'plugin': <MagicMock id='4490204328'>}, arguments = {}

    def run_check(self, plugin, **arguments):
        """Run the check in a single plugin."""
        LOG.debug("Running %r with %r", plugin, arguments)
        try:
            self.processor.keyword_arguments_for(
                plugin["parameters"], arguments
            )
        except AttributeError as ae:
            LOG.error("Plugin requested unknown parameters.")
            raise exceptions.PluginRequestedUnknownParameters(
>               plugin=plugin, exception=ae
            )
E           flake8.exceptions.PluginRequestedUnknownParameters: <unprintable PluginRequestedUnknownParameters object>

src/flake8/checker.py:439: PluginRequestedUnknownParameters
--------------------------------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------------------------------
checker.py                 431 DEBUG    Running {'parameters': [{'ast': True}], 'plugin': <MagicMock id='4490204328'>} with {}
checker.py                 437 ERROR    Plugin requested unknown parameters.
```
Sorry for just passing this on but I really can't make sense of it. How come `self` is `NoneType` when it's the FileChecker?
```py
(Pdb) l
429         def run_check(self, plugin, **arguments):
430             """Run the check in a single plugin."""
431             LOG.debug("Running %r with %r", plugin, arguments)
432             try:
433                 import pdb; pdb.set_trace()
434  ->             self.processor.keyword_arguments_for(
435                     plugin["parameters"], arguments
436                 )
437             except AttributeError as ae:
438                 LOG.error("Plugin requested unknown parameters.")
439                 raise exceptions.PluginRequestedUnknownParameters(
(Pdb) type(self)
<class 'flake8.checker.FileChecker'>
(Pdb) next
AttributeError: 'NoneType' object has no attribute 'keyword_arguments_for'
> /Users/tomer/Forks/flake8/src/flake8/checker.py(434)run_check()
```

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 12, 2019, 16:28

`self.processor` seems to only get set up if the file exists (via `._make_processor`) -- can probably use `tmpdir` fixture and make a real `foobar.py` file

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 13, 2019, 11:25

added 1 commit

<ul><li>494bec00 - Write a test checking plugin failure exception handling</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=31340935&start_sha=1f260fac1d30f6a6dd71a3ff0c18f335540df579)

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 13, 2019, 11:26

added 1 commit

<ul><li>3dff93df - Write a test checking plugin failure exception handling</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=31340950&start_sha=494bec003c8be70d184698b412e1703d0dc5a895)

**Comment by asottile** [comment]
In GitLab by @codecov on Jan 13, 2019, 11:30

# [Codecov](https://codecov.io/gl/pycqa/flake8/pull/275?src=pr&el=h1) Report
> Merging [#275](https://codecov.io/gl/pycqa/flake8/pull/275?src=pr&el=desc) into [master](https://codecov.io/gl/pycqa/flake8/commit/f70ae10c2ba191bef78c723a3404c27d84f13de8?src=pr&el=desc) will **increase** coverage by `0.29%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gl/pycqa/flake8/pull/275/graphs/tree.svg?width=650&token=QL6TVnaXih&height=150&src=pr)](https://codecov.io/gl/pycqa/flake8/pull/275?src=pr&el=tree)

```diff
@@            Coverage Diff            @@
##           master    #275      +/-   ##
=========================================
+ Coverage    71.9%   72.2%   +0.29%     
=========================================
  Files          25      25              
  Lines        2367    2367              
  Branches      393     393              
=========================================
+ Hits         1702    1709       +7     
+ Misses        584     577       -7     
  Partials       81      81
```


| [Impacted Files](https://codecov.io/gl/pycqa/flake8/pull/275?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/flake8/checker.py](https://codecov.io/gl/pycqa/flake8/pull/275/diff?src=pr&el=tree#diff-c3JjL2ZsYWtlOC9jaGVja2VyLnB5) | `57.18% <ø> (+0.95%)` | :arrow_up: |
| [src/flake8/exceptions.py](https://codecov.io/gl/pycqa/flake8/pull/275/diff?src=pr&el=tree#diff-c3JjL2ZsYWtlOC9leGNlcHRpb25zLnB5) | `67.21% <0%> (+6.55%)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gl/pycqa/flake8/pull/275?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gl/pycqa/flake8/pull/275?src=pr&el=footer). Last update [f70ae10...3579e4e](https://codecov.io/gl/pycqa/flake8/pull/275?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 13, 2019, 11:41

added 1 commit

<ul><li>80b47006 - Fix dogfood</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=31341155&start_sha=3dff93df6b6bf6b7fdae8c1fab3d258588eb4109)

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 13, 2019, 11:44

added 1 commit

<ul><li>ef42d743 - Fix python2 unicode test error</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=31341197&start_sha=80b4700619e9afaff5f592cb4826137fd89a5a56)

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 13, 2019, 12:01

added 1 commit

<ul><li>dd94ee7b - Fix python2 unicode test error</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=31341486&start_sha=ef42d74353addc6e2fd5b7d87f1998619d76f915)

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 13, 2019, 12:03

added 1 commit

<ul><li>1df01c42 - Fix python2 unicode test error</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=31341512&start_sha=dd94ee7bcc66d19b18d94499c2066b11a232c179)

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 13, 2019, 12:04

added 1 commit

<ul><li>a7ef7865 - Fix python2 unicode test error</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=31341536&start_sha=1df01c422105a3dc9405cc2d3699ed729bfd3787)

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 13, 2019, 12:05

added 1 commit

<ul><li>eab56c55 - Fix python2 unicode test error</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=31341543&start_sha=a7ef78651b2f9995eb6b93e8d9a59bed509fbdcd)

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 13, 2019, 12:05

added 1 commit

<ul><li>134a05d2 - Fix python2 unicode test error</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=31341557&start_sha=eab56c556c5c0342edcc1eef5c945841ab969629)

**Comment by asottile** [comment]
In GitLab by @Tadaboody on Jan 13, 2019, 12:14

added 1 commit

<ul><li>2b078e3e - Refactor file processor tests to use fixture</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=31341674&start_sha=134a05d26c0f5002138453ddd2735d3c549cd9be)

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 30, 2019, 08:30

mentioned in merge request !291

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 30, 2019, 08:59

added 41 commits

<ul><li>2b078e3e...f70ae10c - 35 commits from branch <code>pycqa:master</code></li><li>c68e7a8d - Fix plugin exception reporting keyerror typo</li><li>b64421c5 - Write a test checking plugin failure exception handling</li><li>bdf3ce91 - Fix dogfood</li><li>92a03768 - Fix python2 unicode test error</li><li>a4f5447f - Refactor file processor tests to use fixture</li><li>3579e4e1 - Fix linters</li></ul>

[Compare with previous version](https://gitlab.com/pycqa/flake8/merge_requests/275/diffs?diff_id=32786888&start_sha=2b078e3e4ac09f71f111c4fed0f60cd136115ade)

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 30, 2019, 09:01

approved this merge request

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 30, 2019, 09:01

enabled an automatic merge when the pipeline for 3579e4e198a44224097c03c68ea9463d0af48bf0 succeeds

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 30, 2019, 09:03

merged

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 30, 2019, 09:03

mentioned in commit 5707e6e3af144b8fc0407e19032f0a5674ba8d0f

**Comment by asottile** [comment]
In GitLab by @asottile on Jan 30, 2019, 12:53

This has been released in 3.7.2 -- thanks again for the PR :tada:
